### DISCUSSION TITLE
Fast message flow cause Client ID collision

### DISCUSSION THREAD

**Discussion by edaspb** [nanomq/nanomq#1921] [Q&A] [answered]
Hi everyone,

I'm not sure if it is a bug or expected behavior.

I'm testing mqtt_client from esp-idf framework with pppos (point-to-point over serial) network stack and it gives the following error on the server side in one specific scenario:

```
[17] ERROR /nanomq/nng/src/core/pipe.c:518 nni_pipe_set_pid: Client ID collision or set ID failed!
 [11] WARN  /nanomq/nng/src/sp/transport/mqtt/broker_tcp.c:679 tcptran_pipe_recv_cb: nni aio recv error!! Object closed
[11] WARN  /nanomq/nng/src/sp/transport/mqtt/broker_tcp.c:964 tcptran_pipe_recv_cb: tcptran_pipe_recv_cb: recv_error rv: 139
 [14] ERROR /nanomq/nng/src/core/pipe.c:518 nni_pipe_set_pid: Client ID collision or set ID failed!
[12] WARN  /nanomq/nng/src/sp/transport/mqtt/broker_tcp.c:679 tcptran_pipe_recv_cb: nni aio recv error!! Object closed
12] WARN  /nanomq/nng/src/sp/transport/mqtt/broker_tcp.c:964 tcptran_pipe_recv_cb: tcptran_pipe_recv_cb: recv_error rv: 139
```
In this scenario mqtt_client sends two or more small messages one by one (I use blocking function esp_mqtt_client_publish) in a relatively fast rate.

I didn't see that with another popular libs like PubSubClient.
Is this a bug or this error is expected and I need to limit the message flow rate?
nanomq 0.22.9

**Comment by JaylinYu** [comment, marked as answer]
Those lines of logs are not correlated.
`Client ID collision or set ID failed!` says you have 2 client using the same clientid. Which is probably caused by reconnect of esp-idf board due to instable wifi.

By publishing msg alone will NOT hit this log.

**Reply by edaspb** [reply]
Hi @JaylinYu ! Thanks for your reply.

That behavior can be 100% reproduced with fast flow message rate. Let's say in happens when esp32 (micro controller) sends two one byte messages one-by-one.  No network issue observes. And esp32 is the unique client for the nanomq instance.
I will investigate further.

**Reply by JaylinYu** [reply]
Fine. However log says very clear there is client disconnecting, no matter it is done by 1 or 2 clients

**Comment by edaspb** [comment]
I think I found out the cause, it was big session keepalive time. It was configured to 600s.
The client disconnecting was related to another [issue](https://github.com/nanomq/nanomq/issues/1922).
Thanks @JaylinYu for you time, appreciate your attention.
BTW it was discovered with web api, very useful feature.
